homeassistant/components/homekit_controller/number.py

```diff
@@ -110,3 +90,98 @@ async def async_set_value(self, value: float):
                 self._char.type: value,
             }
         )
+
+
+class HomeKitEcobeeFanModeNumber(CharacteristicEntity, NumberEntity):
+    """Representation of a Number control for Ecobee Fan Mode request."""
+
+    _attr_name = "Fan Mode"
+
+    def get_characteristic_types(self):
+        """Define the homekit characteristics the entity is tracking."""
+        return [self._char.type]
+
+    @property
+    def name(self) -> str:
+        """Return the name of the device if any."""
+        prefix = ""
+        if name := super().name:
+            prefix = name
+        return f"{prefix} {self._attr_name}"
+
+    @property
+    def min_value(self) -> float:
+        """Return the minimum value."""
+        return self._char.minValue
+
+    @property
+    def max_value(self) -> float:
+        """Return the maximum value."""
+        return self._char.maxValue
+
+    @property
+    def step(self) -> float:
+        """Return the increment/decrement step."""
+        return self._char.minStep
+
+    @property
+    def value(self) -> float:
+        """Return the current characteristic value."""
+        return self._char.value
+
+    async def async_set_value(self, value: float):
+        """Set the characteristic to this value."""
+        if value > self.min_value:
+            other_value = value - 1
+        else:
+            other_value = self.min_value + 1
+
+        if value != other_value:
+            await self.async_put_characteristics(
+                {
+                    self._char.type: other_value,
+                }
+            )
+
+        await self.async_put_characteristics(
+            {
+                self._char.type: value,
+            }
+        )
+
+
+NUMBER_ENTITIES: dict[str, NumberEntityDescription] = {
+    CharacteristicsTypes.Vendor.VOCOLINC_HUMIDIFIER_SPRAY_LEVEL: NumberEntityDescription(
+        key=CharacteristicsTypes.Vendor.VOCOLINC_HUMIDIFIER_SPRAY_LEVEL,
+        name="Spray Quantity",
+        icon="mdi:water",
+        entity_category=EntityCategory.CONFIG,
+    ),
+    CharacteristicsTypes.Vendor.EVE_DEGREE_ELEVATION: NumberEntityDescription(
+        key=CharacteristicsTypes.Vendor.EVE_DEGREE_ELEVATION,
+        name="Elevation",
+        icon="mdi:elevation-rise",
+        entity_category=EntityCategory.CONFIG,
+    ),
+    CharacteristicsTypes.Vendor.AQARA_GATEWAY_VOLUME: NumberEntityDescription(
+        key=CharacteristicsTypes.Vendor.AQARA_GATEWAY_VOLUME,
+        name="Volume",
+        icon="mdi:volume-high",
+        entity_category=EntityCategory.CONFIG,
+    ),
+    CharacteristicsTypes.Vendor.AQARA_E1_GATEWAY_VOLUME: NumberEntityDescription(
+        key=CharacteristicsTypes.Vendor.AQARA_E1_GATEWAY_VOLUME,
+        name="Volume",
+        icon="mdi:volume-high",
+        entity_category=EntityCategory.CONFIG,
+    ),
+    CharacteristicsTypes.Vendor.ECOBEE_SET_HOLD_SCHEDULE: NumberEntityDescription(
```

Comment:
NumberEntity is meant for r/w entities, but according to the schema we have in the repo ECOBEE_SET_HOLD_SCHEDULE is w/o. Because of the limitation of these characteristics I think we need a custom entity that writes to `ECOBEE_SET_HOLD_SCHEDULE` but reads from `ECOBEE_CURRENT_MODE`.

I think if we are doing that it might be better to make a select entity (https://developers.home-assistant.io/docs/core/entity/select).

Maybe drop preset mode from this PR, and add it back in a new one? Then we can still hopefully get the others merged.

Reply:
PR: https://github.com/home-assistant/core/pull/65032